### PR TITLE
gh-124855: Don't allow the JIT and perf support to be active at the same time

### PR DESCRIPTION
**Pull request by pablogsal**
* Issue: gh-124855

### CODE REVIEW

File: Python/sysmodule.c [hunk 1]
```diff
@@ -2287,6 +2287,14 @@ sys_activate_stack_trampoline_impl(PyObject *module, const char *backend)
 /*[clinic end generated code: output=5783cdeb51874b43 input=a12df928758a82b4]*/
 {
 #ifdef PY_HAVE_PERF_TRAMPOLINE
+#ifdef _Py_TIER2
```
Comment (Fidget-Spinner) [inline Python/sysmodule.c]: The macro you want is `_Py_JIT`. `_Py_TIER2` also includes the tier 2 interpreter which is compatible with the trampolines.

File: Python/sysmodule.c [hunk 2]
```diff
@@ -2287,6 +2287,14 @@ sys_activate_stack_trampoline_impl(PyObject *module, const char *backend)
...
+    _PyOptimizerObject* optimizer = _Py_GetOptimizer();
+    if (optimizer != NULL) {
+        PyErr_SetString(PyExc_ValueError, "Cannot activate the perf trampiline if the JIT is active");
```
Comment (Fidget-Spinner) [inline Python/sysmodule.c]:
```suggestion
        PyErr_SetString(PyExc_ValueError, "Cannot activate the perf trampoline if the JIT is active");
```

File: Python/sysmodule.c (hunk 1 above)
Comment (pablogsal) [inline Python/sysmodule.c]:
```suggestion
#ifdef _Py_JIT
```

Comment (pablogsal) [reply Python/sysmodule.c]: Thanks for the catch @Fidget-Spinner !

File: Misc/NEWS.d/next/Core_and_Builtins/2024-10-01-17-31-32.gh-issue-124855.sdsv_H.rst [hunk 3]
```diff
@@ -0,0 +1,2 @@
+Don't allow the JIT and perf support to be active at the same time. Patch by
```
Comment (rruuaanng) [inline Misc/NEWS.d/next/Core_and_Builtins/2024-10-01-17-31-32.gh-issue-124855.sdsv_H.rst:1]: Maybe you should remove the author name, as it doesn't mean anything to the user.

Comment (pablogsal) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-10-01-17-31-32.gh-issue-124855.sdsv_H.rst:1]: I (as many others) have been doing this for 7 years to mark the contributions :)

Comment (pablogsal) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-10-01-17-31-32.gh-issue-124855.sdsv_H.rst:1]: You can search for "patch by" in https://docs.python.org/3.13/whatsnew/changelog.html#changelog to see how common this is

Comment (diegorusso) [reply Misc/NEWS.d/next/Core_and_Builtins/2024-10-01-17-31-32.gh-issue-124855.sdsv_H.rst:1]: TBH I find it useful as by reading the changelog I have an idea who did what and I don't have to go through the endless PRs on GitHub.

**Comment by diegorusso** [comment]
Do we need to update the documentation here? https://docs.python.org/3.14/library/sys.html#sys.activate_stack_trampoline

**Comment by pablogsal** [comment]
Good call!

File: Python/sysmodule.c [hunk 4]
```diff
@@ -2287,6 +2287,14 @@ sys_activate_stack_trampoline_impl(PyObject *module, const char *backend)
 /*[clinic end generated code: output=5783cdeb51874b43 input=a12df928758a82b4]*/
 {
 #ifdef PY_HAVE_PERF_TRAMPOLINE
+#ifdef _Py_JIT
+    _PyOptimizerObject* optimizer = _Py_GetOptimizer();
+    if (optimizer != NULL) {
+        PyErr_SetString(PyExc_ValueError, "Cannot activate the perf trampoline if the JIT is active");
+        return NULL;
+    }
+#endif
```
Comment (Fidget-Spinner) [inline Python/sysmodule.c:2296]: You're missing a `Py_DECREF(optimizer)` here I think

File: Python/sysmodule.c [hunk 5]
```diff
@@ -2287,6 +2287,14 @@ sys_activate_stack_trampoline_impl(PyObject *module, const char *backend)
 /*[clinic end generated code: output=5783cdeb51874b43 input=a12df928758a82b4]*/
 {
 #ifdef PY_HAVE_PERF_TRAMPOLINE
+#ifdef _Py_JIT
+    _PyOptimizerObject* optimizer = _Py_GetOptimizer();
+    if (optimizer != NULL) {
+        PyErr_SetString(PyExc_ValueError, "Cannot activate the perf trampoline if the JIT is active");
```
Comment (Fidget-Spinner) [inline Python/sysmodule.c:2293]: You're forgetting a `Py_DECREF(optimizer)` I think.

Comment (Fidget-Spinner) [reply Python/sysmodule.c:2293]: @pablogsal I'm pretty certain there's a refleak here :(

Comment (pablogsal) [reply Python/sysmodule.c:2293]: On it, sorry I had this on auto-land and I missed your comment :(

Comment (pablogsal) [reply Python/sysmodule.c:2293]: https://github.com/python/cpython/pull/126163

**Pull request merged**